tpm/tpm.go

```diff
@@ -0,0 +1,199 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tpm implements TPM2 API to seal and Unseal data.
+package tpm
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var (
+	// Default EK template defined in:
+	// https://trustedcomputinggroup.org/wp-content/uploads/Credential_Profile_EK_V2.0_R14_published.pdf
+	// Shared SRK template based off of EK template and specified in:
+	// https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
+	srkTemplate = tpm2.Public{
+		Type:       tpm2.AlgRSA,
+		NameAlg:    tpm2.AlgSHA256,
+		Attributes: tpm2.FlagFixedTPM | tpm2.FlagFixedParent | tpm2.FlagSensitiveDataOrigin | tpm2.FlagUserWithAuth | tpm2.FlagRestricted | tpm2.FlagDecrypt | tpm2.FlagNoDA,
+		AuthPolicy: nil,
+		RSAParameters: &tpm2.RSAParams{
+			Symmetric: &tpm2.SymScheme{
+				Alg:     tpm2.AlgAES,
+				KeyBits: 128,
+				Mode:    tpm2.AlgCFB,
+			},
+			KeyBits:    2048,
+			Exponent:   0,
+			ModulusRaw: make([]byte, 256),
+		},
+	}
+)
+
+func Seal(tpmPath string, pcr int, srkPassword, objectPassword string, dataToSeal []byte) (privateArea, publicArea []byte, retErr error) {
+	// Open the TPM
+	rwc, err := tpm2.OpenTPM(tpmPath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("can't open TPM %q: %v", tpmPath, err)
+	}
+	defer func() {
+		if err := rwc.Close(); err != nil {
+			retErr = fmt.Errorf("%v\ncan't close TPM %q: %v", retErr, tpmPath, err)
+		}
+	}()
+
+	// Create the parent key against which to seal the data
+	srkHandle, _, err := tpm2.CreatePrimary(rwc, tpm2.HandleOwner, tpm2.PCRSelection{}, "", srkPassword, srkTemplate)
+	if err != nil {
+		return nil, nil, fmt.Errorf("can't create primary key: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, srkHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush SRK handle %q: %v", retErr, srkHandle, err)
+		}
+	}()
+	glog.Infof("Created parent key with handle: 0x%x\n", srkHandle)
+
+	// Note the value of the pcr against which we will seal the data
+	pcrVal, err := tpm2.ReadPCR(rwc, pcr, tpm2.AlgSHA256)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to read PCR: %v", err)
+	}
+	glog.Infof("PCR %v value: 0x%x\n", pcr, pcrVal)
+
+	// Get the authorization policy that will protect the data to be sealed
+	sessHandle, policy, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to get policy: %v", err)
+	}
+	if err := tpm2.FlushContext(rwc, sessHandle); err != nil {
+		return nil, nil, fmt.Errorf("unable to flush session: %v", err)
+	}
+	glog.Infof("Created authorization policy: 0x%x\n", policy)
+
+	// Seal the data to the parent key and the policy
+	privateArea, publicArea, retErr = tpm2.Seal(rwc, srkHandle, srkPassword, objectPassword, policy, dataToSeal)
```

Comment:
I believe size limit for `dataToSeal` is very small.
Have you tried this with a real key of some kind?

Reply:
It is enough to seal the client_secret section - I should be able to construct the rest of the token.
Will play with this more.
Not sure if I need to seal the refresh token as well, hopefully not.